src/pal/src/sharedmemory/sharedmemory.cpp

```diff
@@ -0,0 +1,1114 @@
+
+            default:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+    }
+}
+
+void SharedMemoryHelpers::CloseFile(int fileDescriptor)
+{
+    _ASSERTE(fileDescriptor != -1);
+
+    int closeResult;
+    do
+    {
+        closeResult = close(fileDescriptor);
+    } while (closeResult == EINTR);
+}
+
+SIZE_T SharedMemoryHelpers::GetFileSize(int fileDescriptor)
+{
+    _ASSERTE(fileDescriptor != -1);
+
+    off_t endOffset = lseek(fileDescriptor, 0, SEEK_END);
+    if (endOffset == static_cast<off_t>(-1) ||
+        lseek(fileDescriptor, 0, SEEK_SET) == static_cast<off_t>(-1))
+    {
+        throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+    }
+    return endOffset;
+}
+
+void SharedMemoryHelpers::SetFileSize(int fileDescriptor, SIZE_T byteCount)
+{
+    _ASSERTE(fileDescriptor != -1);
+    _ASSERTE(static_cast<off_t>(byteCount) == byteCount);
+
+    while (true)
+    {
+        int ftruncateResult = ftruncate(fileDescriptor, static_cast<off_t>(byteCount));
+        if (ftruncateResult == 0)
+        {
+            break;
+        }
+        if (errno != EINTR)
+        {
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+        }
+    }
+}
+
+void *SharedMemoryHelpers::MemoryMapFile(int fileDescriptor, SIZE_T byteCount)
+{
+    _ASSERTE(fileDescriptor != -1);
+    _ASSERTE(byteCount > sizeof(SharedMemorySharedDataHeader));
+    _ASSERTE(AlignDown(byteCount, PAGE_SIZE) == byteCount);
+
+    void *sharedMemoryBuffer = mmap(nullptr, byteCount, PROT_READ | PROT_WRITE, MAP_SHARED, fileDescriptor, 0);
+    if (sharedMemoryBuffer != MAP_FAILED)
+    {
+        return sharedMemoryBuffer;
+    }
+    switch (errno)
+    {
+        case ENFILE:
+        case ENOMEM:
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::OutOfMemory));
+
+        default:
+            throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::IO));
+    }
+}
+
+bool SharedMemoryHelpers::TryAcquireFileLock(int fileDescriptor, int operation)
+{
+    // A file lock is acquired once per file descriptor, so the caller will need to synchronize threads of this process
+
+    _ASSERTE(fileDescriptor != -1);
+    _ASSERTE(!(operation & LOCK_UN));
+
+    while (true)
+    {
+        if (flock(fileDescriptor, operation) == 0)
+        {
+            return true;
+        }
+
+        int flockError = errno;
+        switch (flockError)
+        {
+            case EWOULDBLOCK:
+                return false;
+
+            case EINTR:
+                continue;
+
+            default:
+                throw SharedMemoryException(static_cast<DWORD>(SharedMemoryError::OutOfMemory));
+        }
+    }
+}
+
+void SharedMemoryHelpers::ReleaseFileLock(int fileDescriptor)
+{
+    _ASSERTE(fileDescriptor != -1);
+
+    int flockResult = flock(fileDescriptor, LOCK_UN);
```

Comment:
I believe the system call may only be interrupted during a wait for a lock:

"EINTR  While waiting to acquire a lock, the call was interrupted by delivery of a signal caught by a handler; see signal(7)."

Doesn't hurt to check for it though, added a check for EINTR anyway.